server/hotrod/src/main/scala/org/infinispan/server/hotrod/HotRodServer.scala

```diff
@@ -302,26 +303,44 @@ class HotRodServer extends AbstractProtocolServer("HotRod") with Log {
 
    @Listener(sync = false)
    class ReAddMyAddressListener(addressCache: AddressCache, clusterAddress: Address, address: ServerAddress) {
-      // We rely on the fact that even with partition handling disabled, the data in the minority
-      // partition is wiped on a merge.
-      // We also listen for removal events, in case we receive (and apply) the other partition's remove
-      // command after the merge.
-      @CacheEntryInvalidated @CacheEntryRemoved
-      def addressRemoved(event: CacheEntryEvent[Address, ServerAddress]): Unit = {
+      @TopologyChanged
+      def topologyChanged(event: TopologyChangedEvent[Address, ServerAddress]): Unit = {
          if (event.isPre)
             return
 
-         if (event.getKey.equals(clusterAddress)) {
-            // Our address just got invalidated out of the cache, most likely because of a merge
-            // Put it back asynchronously, so we don't block the topology update
-            log.debugf("Re-adding %s to the topology cache", clusterAddress)
-            addressCache.put(clusterAddress, address)
+         val distributedExecutor = new DefaultExecutorService(addressCache)
+         var success = false
+         while (!success) {
+            try {
+               val futures = distributedExecutor.submitEverywhere(new CheckAddressTask(clusterAddress, address))
+               val everybodyHasIt = futures.forall(_.get())
```

Comment:
If you look at the `DistributedTaskPart.innerGet()` implementation, it's not actually waiting forever, because the task has a default timeout equal to `sync.replTimeout`. I could use the same timeout here, but I'm not sure it would help.